### ISSUE TITLE
Discrepancy in type overlap checking between pattern matching and if-equal when working with IntEnums

### ISSUE THREAD

**Issue by Wizzerinus** [DetachHead/basedpyright#1701]
### Description

So here's the playground link:

https://basedpyright.com/?pyrightVersion=1.31.4&typeCheckingMode=all&reportUnusedParameter=false&code=GYJw9gtgBApgdgV2gSwgBzCALlAknLAUUQgBooBDBLMAWAChRIosBPNZOAcylQ2yjBOFADYNx9AAJC4ohgGMRFAM7KoAMTBgAFPiIkAlAC4GUM5SgBeStR0HT5gEZWbNbffoSGAExjAoAB7uJvTmFtbammAAdBTkAEweDmbI-hRW1lGxIWFhaCrKyVAQFFjyABaUOblm8iowGlrZRTVm%2Baot5nXKDQD61a3m7cpAA

On line 15, Basedpyright reports that a variable of type `tuple[Foo, int]` can never be equal to `Foo.a`. This makes sense to me, since `Foo` is a final class, so the only way to make a `Foo` is to use one of its own variants, which `tuple[Foo, int]` is not.

However, on line 17, Basedpyright gladly accepts matching against `Foo.a`, despite it having essentially the same semantics.

This issue is only meaningful in Basedpyright, since Pyright doesn't report reportUnnecessaryComparison in this case. (And I think it should! But I guess this is a based-specific feature)

**Comment by DetachHead** [comment]
interesting. also if we modify the example to compare an `int` to an `IntEnum`, it reports the error in the `match` case but not the `if` statement, when it shouldn't report an error at all:

Code sample in [basedpyright playground](https://basedpyright.com/?pyrightVersion=1.31.4&typeCheckingMode=all&reportUnusedParameter=false&code=GYJw9gtgBApgdgV2gSwgBzCALlAknLAUUQgBooBDBLMAWAChRIosBPNZOAcylQ2yjBOFADYNx9AAJC4ohgGMRFAM7KoAMTBgAFPiIkAlAC4GUM5SgBeStR0HT5gEZWbNbffoSGAExjAoAB7uJvTmFtYAjBJhyP4UVtaaYAB0FCFhYWgqyg5mEBRY8gAWlOkZZvIqMBpaqWXlmdm5GZXK1QD69Q1mWapAA)

```python
from enum import IntEnum, auto
from typing import final


@final
class Foo(IntEnum):
    a = auto()
    b = auto()



def x():
    a = 1

    if a == Foo.a: # no error
        pass
    match a:
        case Foo.a: # error
            pass
        case _:
            pass
```
 

> This issue is only meaningful in Basedpyright, since Pyright doesn't report reportUnnecessaryComparison in this case. (And I think it should! But I guess this is a based-specific feature)

[it does happen in pyright](https://pyright-play.net/?strict=true&deprecateTypingAliases=true&enableExperimentalFeatures=true&reportCallInDefaultInitializer=true&reportImplicitStringConcatenation=true&reportImplicitOverride=true&reportMissingSuperCall=true&reportPropertyTypeMismatch=true&reportShadowedImports=true&reportUnnecessaryTypeIgnoreComment=true&reportUninitializedInstanceVariable=true&code=GYJw9gtgBApgdgV2gSwgBzCALlAknLAUUQgBooBDBLMAWAChRIosBPNZOAcylQ2yjBOFADYNx9AAJC4ohgGMRFAM7KoAMTBgAFPiIkAlAC4GUM5SgBeStR0HT5gEZWbNbffoSGAExjAoAB7uJvTmFtbammAAdBTkAEweDmbI-hRW1lGxIWFhaCrKyVAQFFjyABaUOblm8iowGlrZRTVm%2Baot5nXKDQD61a3m7cpAA), it's just that `reportUnnecessaryComparison` is disabled by default in pyright
